### ISSUE TITLE
Umbraco.Tags does not work with DTGE

### ISSUE THREAD

**Issue by enkelmedia** [skttl/umbraco-doc-type-grid-editor#173]
Hi!

We had the need to support tags inside a DTGE for one of our projects, I've debugged the issue so far to see that the "TagsValueConverter" in Umbraco-core is expecting a string. This call inside DocTypeGridEditorHelper (around L.87)

`prop2.SetValue(newValue);`

Will expect the passed value to be of the "correct" DB-type, if not, it will do a .ToString() is a string is what is expected. In our case this will result in the string being something like "System.Linq.Enumerable+WhereSelectEnumerableIterator`2[Newtonsoft.Json.Linq.JToken,System.String]" and not the string-array values ["Foo1","Foo2","Foo3"].

I've seen that this is an issue that has been addressed before: https://github.com/skttl/umbraco-doc-type-grid-editor/issues/43

I'm thinking that our chanses of getting a change Umbraco.Core to facilitate this might be small, but maybe we could provide a way to do proccesing of the stored values based on the property editor type?

On L.66 I've changed the code to something like this:
```

  var newValue = propEditor.GetValueEditor().FromEditor(contentPropData, jProp.Value);

   if (propEditor.Alias == "Umbraco.Tags")
   {
     newValue = JsonConvert.SerializeObject(newValue);
   }
```

This way the call to `prop2.SetValue(newValue);` will contain a valid string and the value passed to the property value converter later on will be correct.

I'm thinking that maybe we could add a list of "DtgeValueProcessors" that can be used to process a value before calling the property value converter? It would could be a simple class that looks like a slim version of a property value converter.

Something like this:

```
 public class UmbracoTagsValueProcessor : IDocTypeGridEditorValueProcessor
    {

        public bool IsProcessorFor(string propertyEditorAlias) => propertyEditorAlias.Equals("Umbraco.Tags");

        public object ProcessValue(object value)
        {
            return JsonConvert.SerializeObject(value);
        }

    }
```

Then making it possible to configure these with a Composer, 

```
public class MyComposer : IUserComposer {
        public void Compose(Composition composition)
        {
            composition.DocTypeGridEditorValueProcessors().Append<UmbracoTagsValueProcessor>();
        }
    }
```

Is this something that you would consider for a PR? In that case I'm open to contribute =D

**Comment by skttl** [comment]
Hi @enkelmedia 

That looks very nice - I'll be happy to accept a PR :)

**Comment by skttl** [comment]
But one thing, that should be mentioned in the docs, is that you don't get the complete feature set of the tag editor when using in DTGE though. I can't really see how Umbraco should be able to find nodes using a TagQuery, where the tags is located inside a grid editor (or Nested Content for that sake).

This is also why the previous issue was closed.

For my own use, I have previously made a custom tag editor for this in stead.

**Comment by andersbrohall** [comment]
In this particular case, we use the Tag Query in DTGE to filter a selection of Content Nodes (with tags). 
: )

**Comment by enkelmedia** [comment]
@skttl Yes, like @andersbrohall is saying, when we use it like this it's for choosing a range of filters to display inside the grid. I the node it self needs to be tagged that would be done with a "regular" Tag editor right on the page. 

Like you say, it's the same thing (I guess) with Nested Content. But it's always good to mention.

So I'll go ahead and create a PR for this then =D

Is there any other property editors that might be suitable for this approach? Where the data passed from into the Property Value Converter is not as the converter expects?

Do we like the naming? "IDocTypeGridEditorValueProcessor" and "FooValueProcessor" etc, or is there any other suggestions?

// m

**Comment by skttl** [comment]
I haven't seen any other. Actually I though the tag problem was only around not being able to use TagQuery for things tagged inside grid :)

I think `IDocTypeGridEditorValueProcessor` and `FooValueProcessor` sounds good.

**Comment by skttl** [comment]
Maybe #70 is in the same alley?

**Comment by enkelmedia** [comment]
I our case we could not render the tags since the value passed to the property value converter was something like "System.Linq.Enumerable+WhereSelectEnumerableIterator`2[Newtonsoft.Json.Linq.JToken,System.String" as the Umbraco-core would just do a .ToString() on any object that is not of the correct type.

See: 
https://github.com/umbraco/Umbraco-CMS/blob/12e88fde423f59eae1dff76b474974382d672fe3/src/Umbraco.Core/Models/Property.cs#L268

.SetValue()-calls into here:

https://github.com/umbraco/Umbraco-CMS/blob/12e88fde423f59eae1dff76b474974382d672fe3/src/Umbraco.Core/Models/PropertyType.cs#L324

If the type passed is not the "ExpectedType" it would try to convert it, in our case into a string which gives the nice "System.Linq.Enumerable+WhereSelectEnumerableIterator`2[Newtonsoft.Json.Linq.JToken,System.String"

https://github.com/umbraco/Umbraco-CMS/blob/12e88fde423f59eae1dff76b474974382d672fe3/src/Umbraco.Core/Models/PropertyType.cs#L337

The PR fixes this by making sure that the type of the object passed into SetValue is in fact a string in the right format.

**Comment by hfloyd** [comment]
Hi @enkelmedia & @skttl , I came across this when I was having a similar issue with Tags in a v7 site. I took a look at your PR code, thinking perhaps I could someone integrate something similar into the v-1 branch, for Umbraco 7 usage... but The use of a Composer makes me wonder if this is possible in the same structure in a v7 codebase, since "Composer" is a v8 innovation. Are either of you familiar wit what would be the equivalent way to "register" the ValueProcessors in a v7 site? If you have any ideas, I am willing to pursue them.

Otherwise, I might just go with the less-fancy example originally provided:

> On L.66 I've changed the code to something like this:
> 
>   var newValue = propEditor.GetValueEditor().FromEditor(contentPropData, jProp.Value);
> 
>    if (propEditor.Alias == "Umbraco.Tags")
>    {
>      newValue = JsonConvert.SerializeObject(newValue);
>    }
> This way the call to prop2.SetValue(newValue); will contain a valid string and the value passed to the property value converter later on will be correct.


Thanks for any ideas.
